### PR TITLE
MM-40527 Upgrade to redux-persist@6.0.0 and set it up properly

### PR DESCRIPTION
**Pull request by hmhealey**
Since we're upgrading from 4.x to 6.0.0, there's a lot that's changed. We also had a number of hacks set up to support redux-persist in the way that we were using it, so I had to unwind those to get the new version working correctly. You can see a few of the more specific changes that I had to make by looking at the commit logs, but I'll add comments below to hopefully explain anything too weird that I had to do.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-40527

#### Release Note
```release-note
Update library used for storing drafts and other data in browser storage
```

### CODE REVIEW

File: store/index.js [hunk 1]
```diff
@@ -1,183 +1,166 @@
...
+}
 
-                // check to see if the logout request was successful
-                store.subscribe(() => {
-                    const state = store.getState();
-                    const basePath = getBasePath(state);
+/**
+ * Migrates state.storage from redux-persist@4 to redux-persist@6
+ */
+function migratePersistedState(store, persistor) {
```
Comment (hmhealey) [inline store/index.js:122]: This part is important so that people don't lose their drafts on upgrade

**Review by koox00 — APPROVED** [review]

**Review by jespino — APPROVED** [review]
LGTM

File: e2e/cypress/integration/messaging/message_draft_persistance_spec.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        // # Reload the page
+        // eslint-disable-next-line cypress/no-unnecessary-waiting
+        cy.wait(500).reload();
+
+        // * Ensure the draft is back in the post textbox
+        cy.get('#post_textbox').should('have.text', testText);
+    });
+
+    it('Persisting a draft in another channel', () => {
```
Comment (jgilliam17) [inline e2e/cypress/integration/messaging/message_draft_persistance_spec.js:51]:
```suggestion
    it('MM-T4640 Persisting a draft in another channel', () => {
```

File: e2e/cypress/integration/messaging/message_draft_persistance_spec.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            should('be.visible');
+
+        // # Switch back to Off-Topic
+        cy.get('#sidebarItem_off-topic').click();
+
+        // * Ensure the draft is back in the post textbox
+        cy.get('#post_textbox').should('have.text', testText);
+    });
+
+    it('Migration of drafts from redux-persist@4.0.0', () => {
```
Comment (jgilliam17) [inline e2e/cypress/integration/messaging/message_draft_persistance_spec.js:89]:
```suggestion
    it('MM-T4641 Migration of drafts from redux-persist@4.0.0', () => {
```

**Review by jgilliam17 — COMMENTED** [review]
Thanks for adding E2Es @hmhealey 
See test IDs as suggested below.
We discussed creating an empty PR + server, adding draft data, followed by upgrading the server after new changes are pushed to verify migration is smooth. 
Should we still do that? 
Added E2Es are all passing - see report [here](https://mattermost-cypress-report.s3.amazonaws.com/109426-6abbf5e-onprem-ent-webapp-pr-9713-mattermost/mm-ee-test6abbf5e/mochawesome.html)
Screen Shot 2022-02-04 at 5 12 58 PM

**Comment by hmhealey** [comment]
@jgilliam17 Right, I forgot that we had discussed doing that. I've set up a test server in https://github.com/mattermost/mattermost-webapp/pull/9762 running master currently, so we can coordinate testing on that

**Comment by jgilliam17** [comment]
@hmhealey Tested on the related PR test server https://mattermost-webapp-pr-9762.test.mattermost.cloud/
Looks good to merge. 
No change, post drafts are persisting after migration - as expected.

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**